### PR TITLE
Created util class for parsing rational type to double and string.

### PR DESCRIPTION
**Pull request by michal1990**

### CODE REVIEW

File: app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    private static final String GEO_MINUTES_CHAR = "\'";
+    private static final String GEO_SECONDS_CHAR = "\'\'";
+    private static final String DEFAULT_DENOMINATOR = "/1,";
+    private static final String DENOMINATOR_CHAR = "/";
+    private static final int GEO_SECONDS_DENOMINATOR_VALUE = 1000000;
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    static final char DEGREE_CHAR = (char) 0x00B0;
+    public static final int SIX_DECIMAL_NUMBER_SCALE = 6;
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java]: This should be on top of class

File: app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+    public static String parseDoubleGPSToGeoGPS(double value) {
+        final int[] geoGPS = convertDoubleGPSValueToGeoGPSValue(value);
+
+        int degrees = geoGPS[0];
+        int minutes = geoGPS[1];
+        int seconds = geoGPS[2];
+
+        return new StringBuilder()
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java:36]: This looks like some duplication of code.
Maybe better is to expose method for that and append missing part from other place?

File: app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+     * @param value of GPS represented by double
+     * @return an int array with geo GPS components:
+     * 0 - degrees value
+     * 1 - minutes value
+     * 2 - seconds value
+     */
+    private static int[] convertDoubleGPSValueToGeoGPSValue(double value) {
+        final int[] result = new int[3];
+        result[0] = (int) value;
+        result[1] = (int) (60 * (value - result[0]));
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java]: Looks like "60", "3600" are used in few places, maybe some constant?

File: app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                .append(exactSecondsValue)
+                .append(DENOMINATOR_CHAR)
+                .append(GEO_SECONDS_DENOMINATOR_VALUE)
+                .toString();
+    }
+
+    private static long getExactSecondsValue(double seconds) {
+        BigDecimal big = new BigDecimal(seconds).setScale(SIX_DECIMAL_NUMBER_SCALE, BigDecimal.ROUND_HALF_DOWN);
+        String bigString = big.toString().replace(".", "");
+        return Long.valueOf(bigString);
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java]: This removes dot form double, why?

File: app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import org.junit.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+
+public class TagUtilTest {
+
+    @Test
+    public void whenDoubleGPSIsGiven_GeoGPSWillBeReturn() {
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java]: please keep our naming convention:
"whenDoubleGPSIsGiven_GeoGPSIsReturned()"

File: app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        //given
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java]: Given

File: app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        double gpsValue = 17.071711;
+
+        //when
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java]: When

File: app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        String result = TagUtil.parseDoubleGPSToGeoGPS(gpsValue);
+
+        //then
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java]: Then

**Review by BartSoftware — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java (hunk 2 above)
Comment (michal1990) [reply app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java:36]: Some parts of this builder are duplicated in the second method but I don't think that we should expose method for that.

File: app/src/main/res/layout/edit_item_view.xml [hunk 9]
```diff
@@ -23,7 +23,7 @@
 
             <TextView
                 android:id="@+id/tag_key"
-                android:layout_width="200dp"
+                android:layout_width="170dp"
```
Comment (kejmil01) [inline app/src/main/res/layout/edit_item_view.xml:26]: I think we need to add ellipsize (end) attribute here.
Also, margins between key and value would be nice.

**Review by kejmil01 — APPROVED** [review]

File: app/src/main/java/com/invisibleteam/goinvisible/model/GeolocationTag.java [hunk 10]
```diff
@@ -1,19 +1,73 @@
...
+    }
+
+    private void formatPositionValue() {
+        String geoLat = TagUtil.parseRationalGPSToGeoGPS(getValue(), latitudeRef);
+        String geoLng = TagUtil.parseRationalGPSToGeoGPS(secondValue, longitudeRef);
+
+        formattedValue = geoLat + "\n" + geoLng;
+    }
+
+    public void setSecondKey(String secondKey) {
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/model/GeolocationTag.java]: VisibleForTesting

File: app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/TagsManager.java [hunk 11]
```diff
@@ -52,16 +60,22 @@ boolean editTags(List<Tag> tags) {
...
+        return result;
+    }
+
+    private boolean isPositionTag(Tag tag) {
+        return tag instanceof GeolocationTag;
+    }
+
+    private void saveTagAttributes(String key, String value) throws IOException {
+        exifInterface.setAttribute(key, value);
+        exifInterface.saveAttributes();
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/mvvm/edition/TagsManager.java]: This should be fired once at end for changing tags

File: app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    public static double parseRationalGPSToDoubleGPS(String rational) {
+        String[] rationalParts = rational.split(",");
+
+        double degrees = parseRationalValueToDouble(rationalParts[0]);
+        double minutes = parseRationalValueToDouble(rationalParts[1]);
+        double seconds = parseRationalValueToDouble(rationalParts[2]);
+
+        double result = degrees + minutes / HOUR_IN_MINUTES + seconds / HOUR_IN_SECONDS;
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java]: There is util for parse minutes for seconds etc. (for ex. TimeUnit.SECONDS.toMillis)

**Review by BartSoftware — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+            return null;
+        }
+
+        double result = parseRationalValueToDouble(rational);
+
+        return String.valueOf(result);
+    }
+
+    private static boolean isRationalValueCorrect(@Nullable String rational) {
+        return rational == null || rational.length() == 0 || !rational.contains("/");
```
Comment (BartSoftware) [inline app/src/main/java/com/invisibleteam/goinvisible/util/TagUtil.java]: ( rational == null || rational.length() == 0 ) -> TextUtils.isEmpty(rational)

File: app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        //Given
+        String rational = "18/5";
+
+        //When
+        double result = TagUtil.parseRationalValueToDouble(rational);
+
+        //Then
+        assertThat(result, is(3.6));
+    }
+
```
Comment (BartSoftware) [inline app/src/test/java/com/invisibleteam/goinvisible/util/TagUtilTest.java:166]: Redundant line

**Review by BartSoftware — CHANGES_REQUESTED** [review]

**Review by BartSoftware — APPROVED** [review]

**Pull request merged**